returnn/tf/layers/basic.py

```diff
@@ -3246,40 +3246,71 @@ class RepeatLayer(_ConcatInputLayer):
     """
     :param LayerBase|int repetitions:
-      number of repetitions for each sequence and position in target axis, [B,T] or [T,B]
+      number of repetitions for each sequence and position in target axis.
+      Can be [B,T] or [T,B] or some subset of that shape
     :param str axis: (dynamic) axis for repetition (currently only time axis is supported)
     """
     from returnn.tf.util.data import DimensionTag
     super(RepeatLayer, self).__init__(**kwargs)
     self.repetitions = repetitions
     if isinstance(self.repetitions, int):
-      repetitions_data = Data.from_tensor(tf.constant(self.repetitions)).copy_add_batch_dim(0)
+      repetitions_data = Data.from_tensor(tf.constant(self.repetitions))
     else:
       assert isinstance(self.repetitions, LayerBase)
       repetitions_data = self.repetitions.output
-    input_data = self.input_data
-    assert self.input_data and self.input_data.have_batch_axis()
-    input_data = input_data.copy_as_batch_major()
-    original_axis = input_data.get_axis_from_description(axis, allow_int=False)
-    input_data = input_data.copy_move_axis(original_axis, 1)
-    repetitions_data = repetitions_data.copy_as_batch_major()
+    input_axis = self.input_data.get_axis_from_description(axis, allow_int=False)
     assert repetitions_data.dtype in ["int32", "int64"]
-    assert repetitions_data.have_batch_axis() and repetitions_data.batch_ndim == 2, (
-      "RepeatLayer only accepts repetitions of the shape (B, None|S) or (None|S, B)")
-    assert input_data.shape[0] == repetitions_data.shape[0], (
-      "Axis mismatch between input (%i) and repetitions (%i)" %
-      (input_data.shape[0], repetitions_data.shape[0]))
+    if repetitions_data.ndim == 0:  # no T dim: add a (un)broadcasted one.
+      axis_dim_tag = self.input_data.get_dim_tag(input_axis)
+      repetitions_data = repetitions_data.copy_add_dim_by_tag(axis_dim_tag, unbroadcast=True)
+    repetitions_axis = repetitions_data.get_axis_from_description(axis, allow_int=False)
+    assert repetitions_data.ndim == 1, "Repetitions %r must only have at most one non-batch axis" % repetitions
+    assert repetitions_data.batch_shape[repetitions_axis] == self.input_data.batch_shape[input_axis], (
+      "Axis mismatch between input (%i) and repetitions (%i)" % (
+        self.input_data.batch_shape[input_axis], repetitions_data.batch_shape[repetitions_axis]))
+
+    assert self.output.have_batch_axis() == (self.input_data.have_batch_axis() or repetitions_data.have_batch_axis())
+
+    def copy_placeholder_with_batch_axis(data, other_batch):
+      """
+      Adds a batch dim by unbroadcasting if our output also has a batch dim,
+      otherwise just expands the dims.
+
+      :param Data data:
+      :param BatchInfo|None other_batch: use this batch info if we should add some
+      :rtype: tf.Tensor
+      :return: the placeholder, with shape [B, T, ...], maybe with added batch dim.
+      """
+      if self.output.have_batch_axis():
```

Comment:
Isn't this always true? `self.output` should come from `self.get_out_data_from_opts()` and there it seems you always add a batch axis.

Reply:
In case `repetitions` is an `int`, there might be no batch.

Reply:
Ah, true.

Reply:
My code in `self.get_out_data_from_opts` was also a bit wrong before (see @albertz's comment). In general, `self.output` has a batch dim exactly if the input or the repetitions have one.